File: judo-sdk/src/main/java/com/judopay/api/DeviceDnaInterceptor.java
package com.judopay.api;

import android.content.Context;
import androidx.annotation.NonNull;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.judopay.devicedna.DeviceDNA;

import java.io.IOException;
import java.util.Map;

import okhttp3.Interceptor;
import okhttp3.MediaType;
import okhttp3.RequestBody;
import okhttp3.Response;
import okio.Buffer;

class DeviceDnaInterceptor implements Interceptor {
    private static final String CLIENT_DETAILS = "clientDetails";
    private final DeviceDNA deviceDna;

    DeviceDnaInterceptor(final Context context) {
        this.deviceDna = new DeviceDNA(context);
    }

    @Override
    public Response intercept(@NonNull final Chain chain) throws IOException {
        okhttp3.Request request = chain.request();

        if (isPost(request) && request.body() != null) {
            JsonElement body = getJsonRequestBody(request.body());

            if (body.isJsonObject()) {
                JsonObject json = body.getAsJsonObject();

                if (json.get(CLIENT_DETAILS) == null) {
                    addClientDetails(json);
                }

                return chain.proceed(request.newBuilder()
                        .post(getJsonRequestBody(json))
                        .build());
            }
        }
        return chain.proceed(request);
    }

    private void addClientDetails(final JsonObject json) {
        Map<String, String> signals = deviceDna.getDeviceDNA();
        JsonObject clientDetailsJson = new JsonObject();

        for (Map.Entry<String, String> entry : signals.entrySet()) {
            clientDetailsJson.addProperty(entry.getKey(), entry.getValue());
        }

        json.add(CLIENT_DETAILS, clientDetailsJson);
    }

    private boolean isPost(final okhttp3.Request request) {
        return request.body() != null && "POST".equals(request.method());
    }

    private RequestBody getJsonRequestBody(final JsonObject json) {
        MediaType mediaType = MediaType.parse("application/json");

        return RequestBody.create(mediaType, json.toString());
    }

    private JsonElement getJsonRequestBody(final RequestBody request) throws IOException {
        final Buffer buffer = new Buffer();

        request.writeTo(buffer);
        String body = buffer.readUtf8();

        JsonParser parser = new JsonParser();
        return parser.parse(body);
    }
}

File: judokit-android/src/main/java/com/judopay/judokit/android/api/interceptor/DeviceDnaInterceptor.kt
package com.judopay.judokit.android.api.interceptor

import android.content.Context
import com.google.gson.Gson
import com.google.gson.JsonObject
import com.google.gson.JsonParseException
import com.google.gson.JsonParser
import com.judopay.judokit.android.api.DeviceDetailsProvider
import okhttp3.Interceptor
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okio.Buffer
import java.io.IOException

private const val DEVICE_DETAILS = "deviceDetails"

internal class DeviceDnaInterceptor(
    context: Context,
    private val deviceDetailsProvider: DeviceDetailsProvider = DeviceDetailsProvider(context),
) : Interceptor {
    private val gson = Gson()

    @Throws(IOException::class)
    @Suppress("ReturnCount")
    override fun intercept(chain: Interceptor.Chain): Response {
        val request = chain.request()
        val body = request.body

        if (request.method != "POST" || body == null || !body.isRewritableJson) {
            return chain.proceed(request)
        }

        val postJson = body.bodyAsJsonObject()

        if (postJson == null || postJson.has(DEVICE_DETAILS)) {
            return chain.proceed(request)
        }

        postJson.add(DEVICE_DETAILS, gson.toJsonTree(deviceDetailsProvider.deviceDetails))

        val requestBody = postJson.toString().toRequestBody(body.contentType())

        return chain.proceed(
            request
                .newBuilder()
                .post(requestBody)
                .build(),
        )
    }

    private fun RequestBody.bodyAsJsonObject(): JsonObject? =
        try {
            val buffer = Buffer().also { writeTo(it) }
            JsonParser.parseString(buffer.readUtf8()).asJsonObject
        } catch (ignore: IOException) {
            null
        } catch (ignore: JsonParseException) {
            null
        } catch (ignore: IllegalStateException) {
            null
        }
}

// A one-shot body cannot be buffered here and then written again when the request is sent.
private val RequestBody.isRewritableJson: Boolean
    get() = !isOneShot() && contentType()?.subtype == "json"
